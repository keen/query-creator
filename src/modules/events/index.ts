import { eventsReducer, initialState } from './reducer';
import {
  setCollectionSchemaLoading,
  setEventsCollections,
  fetchCollectionSchema,
  fetchCollectionSchemaSuccess,
  fetchCollectionSchemaError,
  computeSchemaSuccess,
} from './actions';
import {
  getEventsCollections,
  getCollectionSchema,
  getSchemaLoading,
  getSchemas,
} from './selectors';
import {
  FETCH_COLLECTION_SCHEMA,
  FETCH_COLLECTION_SCHEMA_SUCCESS,
  SCHEMA_COMPUTED,
} from './constants';
import { ReducerState, FetchCollectionSchemaAction } from './types';

export {
  initialState,
  eventsReducer,
  fetchCollectionSchema,
  fetchCollectionSchemaSuccess,
  fetchCollectionSchemaError,
  setEventsCollections,
  setCollectionSchemaLoading,
  getEventsCollections,
  getCollectionSchema,
  getSchemas,
  getSchemaLoading,
  computeSchemaSuccess,
  ReducerState,
  FetchCollectionSchemaAction,
};
export {
  FETCH_COLLECTION_SCHEMA,
  FETCH_COLLECTION_SCHEMA_SUCCESS,
  SCHEMA_COMPUTED,
};
