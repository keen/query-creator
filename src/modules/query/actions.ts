import {
  SERIALIZE_QUERY,
  POST_PROCESSING_FINISHED,
  SET_QUERY,
  SELECT_EVENT_COLLECTION,
  SELECT_ANALYSIS,
  SELECT_TARGET_PROPERTY,
  SET_PERCENTILE,
  SET_TIMEFRAME,
  SET_GROUP_BY,
  SET_ORDER_BY,
  SET_PROPERTY_NAMES,
  SET_LIMIT,
  SET_INTERVAL,
  SET_EXTRACTION_LIMIT,
  SET_FILTERS,
  SELECT_TIMEZONE,
  ADD_FUNNEL_STEP,
  CLONE_FUNNEL_STEP,
  SELECT_FUNNEL_STEP_EVENT_COLLECTION,
  UPDATE_FUNNEL_STEP,
  REMOVE_FUNNEL_STEP,
  SET_FUNNEL_STEPS,
  SET_FUNNEL_STEP_FILTERS,
  CHANGE_FUNNEL_STEPS_ORDER,
  ADD_FUNNEL_STEP_FILTER,
  UPDATE_FUNNEL_STEP_FILTER,
  REMOVE_FUNNEL_STEP_FILTER,
  RESET_EXTRACTION,
  RESET_QUERY,
  ADD_FILTER,
  UPDATE_FILTER,
  REMOVE_FILTER,
  UPDATE_FUNNEL_STEP_TIMEZONE,
} from './constants';

import { QueryActions, ReducerState as Query, InitialQuery } from './types';
import {
  Timezones,
  Timeframe,
  OrderBy,
  FunnelStep,
  ExtractionProperty,
  Filter,
} from '../../types';
import { Analysis } from '../../types';

export const setQuery = (query: Partial<Query>): QueryActions => ({
  type: SET_QUERY,
  payload: {
    query,
  },
});

export const serializeQuery = (query: Partial<InitialQuery>): QueryActions => ({
  type: SERIALIZE_QUERY,
  payload: {
    query,
  },
});

export const postProcessingFinished = (): QueryActions => ({
  type: POST_PROCESSING_FINISHED,
});

export const resetQuery = (): QueryActions => ({
  type: RESET_QUERY,
});

export const addFilter = (id: string): QueryActions => ({
  type: ADD_FILTER,
  payload: {
    id,
  },
});

export const removeFilter = (id: string): QueryActions => ({
  type: REMOVE_FILTER,
  payload: { id },
});

export const updateFilter = (
  id: string,
  filter: Partial<Filter>
): QueryActions => ({
  type: UPDATE_FILTER,
  payload: { id, filter },
});

export const setPropertyNames = (
  propertyNames: ExtractionProperty[]
): QueryActions => ({
  type: SET_PROPERTY_NAMES,
  payload: {
    propertyNames,
  },
});

export const selectEventCollection = (name: string): QueryActions => ({
  type: SELECT_EVENT_COLLECTION,
  payload: {
    name,
  },
});

export const selectAnalysis = (analysis: Analysis): QueryActions => ({
  type: SELECT_ANALYSIS,
  payload: {
    analysis,
  },
});

export const selectTargetProperty = (property: string): QueryActions => ({
  type: SELECT_TARGET_PROPERTY,
  payload: {
    property,
  },
});

export const selectTimezone = (timezone: Timezones): QueryActions => ({
  type: SELECT_TIMEZONE,
  payload: {
    timezone,
  },
});

export const setInterval = (interval: string): QueryActions => ({
  type: SET_INTERVAL,
  payload: {
    interval,
  },
});

export const setPercentile = (percentile: number): QueryActions => ({
  type: SET_PERCENTILE,
  payload: {
    percentile,
  },
});

export const setGroupBy = (groupBy?: string | string[]): QueryActions => ({
  type: SET_GROUP_BY,
  payload: {
    groupBy,
  },
});

export const setOrderBy = (orderBy?: OrderBy[]): QueryActions => ({
  type: SET_ORDER_BY,
  payload: {
    orderBy,
  },
});

export const setLimit = (limit?: number): QueryActions => ({
  type: SET_LIMIT,
  payload: {
    limit,
  },
});

export const setExtractionLimit = (limit?: number): QueryActions => ({
  type: SET_EXTRACTION_LIMIT,
  payload: {
    limit,
  },
});

export const resetExtraction = (): QueryActions => ({
  type: RESET_EXTRACTION,
});

export const setTimeframe = (timeframe: Timeframe): QueryActions => ({
  type: SET_TIMEFRAME,
  payload: {
    timeframe,
  },
});

export const setFilters = (filters: Filter[]): QueryActions => ({
  type: SET_FILTERS,
  payload: {
    filters,
  },
});

export const addFunnelStep = (id: string): QueryActions => ({
  type: ADD_FUNNEL_STEP,
  payload: {
    id,
  },
});

export const cloneFunnelStep = (
  cloneId: string,
  newId: string
): QueryActions => ({
  type: CLONE_FUNNEL_STEP,
  payload: {
    cloneId,
    newId,
  },
});
export const selectFunnelStepCollection = (name: string): QueryActions => ({
  type: SELECT_FUNNEL_STEP_EVENT_COLLECTION,
  payload: {
    name,
  },
});

export const updateFunnelStep = (
  stepId: string,
  properties: Partial<FunnelStep>
): QueryActions => ({
  type: UPDATE_FUNNEL_STEP,
  payload: {
    stepId,
    properties,
  },
});

export const changeFunnelStepsOrder = (steps: FunnelStep[]): QueryActions => ({
  type: CHANGE_FUNNEL_STEPS_ORDER,
  payload: { steps },
});

export const setFunnelSteps = (steps: FunnelStep[]): QueryActions => ({
  type: SET_FUNNEL_STEPS,
  payload: { steps },
});

export const setFunnelStepFilters = (
  stepId: string,
  filters: Filter[]
): QueryActions => ({
  type: SET_FUNNEL_STEP_FILTERS,
  payload: {
    stepId,
    filters,
  },
});

export const removeFunnelStep = (stepId: string): QueryActions => ({
  type: REMOVE_FUNNEL_STEP,
  payload: {
    stepId,
  },
});

export const addFunnelStepFilter = (
  stepId: string,
  filterId: string
): QueryActions => ({
  type: ADD_FUNNEL_STEP_FILTER,
  payload: {
    stepId,
    filterId,
  },
});

export const updateFunnelStepFilter = (
  stepId: string,
  filterId: string,
  properties: Filter
): QueryActions => ({
  type: UPDATE_FUNNEL_STEP_FILTER,
  payload: {
    stepId,
    filterId,
    properties,
  },
});

export const removeFunnelStepFilter = (
  stepId: string,
  filterId: string
): QueryActions => ({
  type: REMOVE_FUNNEL_STEP_FILTER,
  payload: {
    stepId,
    filterId,
  },
});

export const updateFunnelStepTimezone = (
  stepId: string,
  timezone: Timezones
): QueryActions => ({
  type: UPDATE_FUNNEL_STEP_TIMEZONE,
  payload: {
    stepId,
    timezone,
  },
});
