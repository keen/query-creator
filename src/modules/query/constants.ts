import { FunnelStep } from '../../types';

export const SET_QUERY = '@query/SET_QUERY';
export const SERIALIZE_QUERY = '@query/SERIALIZE_QUERY';
export const POST_PROCESSING_FINISHED = '@query/POST_PROCESSING_FINISHED';
export const RESET_QUERY = '@query/RESET_QUERY';
export const SELECT_EVENT_COLLECTION = '@query/SELECT_EVENT_COLLECTION';
export const SELECT_ANALYSIS = '@query/SELECT_ANALYSIS';
export const SELECT_TARGET_PROPERTY = '@query/SELECT_TARGET_PROPERTY';
export const SET_PERCENTILE = '@query/SET_PERCENTILE';
export const SET_TIMEFRAME = '@query/SET_TIMEFRAME';
export const SET_GROUP_BY = '@query/SET_GROUP_BY';
export const SET_INTERVAL = '@query/SET_INTERVAL';
export const SET_ORDER_BY = '@query/SET_ORDER_BY';
export const SET_LIMIT = '@query/SET_LIMIT';
export const RESET_EXTRACTION = '@query/RESET_EXTRACTION';
export const SET_EXTRACTION_LIMIT = '@query/SET_EXTRACTION_LIMIT';
export const SET_PROPERTY_NAMES = '@query/SET_PROPERTY_NAMES';
export const SET_FILTERS = '@query/SET_FILTERS';
export const SELECT_TIMEZONE = '@query/SELECT_TIMEZONE';
export const ADD_FUNNEL_STEP = '@query/ADD_FUNNEL_STEP';
export const SELECT_FUNNEL_STEP_EVENT_COLLECTION =
  '@query/SELECT_FUNNEL_STEP_EVENT_COLLECTION';
export const REMOVE_FUNNEL_STEP = '@query/REMOVE_FUNNEL_STEP';
export const UPDATE_FUNNEL_STEP = '@query/UPDATE_FUNNEL_STEP';
export const CLONE_FUNNEL_STEP = '@query/CLONE_FUNNEL_STEP';
export const SET_FUNNEL_STEPS = '@query/SET_FUNNEL_STEPS';
export const SET_FUNNEL_STEP_FILTERS = '@query/SET_FUNNEL_STEP_FILTERS';
export const CHANGE_FUNNEL_STEPS_ORDER = '@query/CHANGE_FUNNEL_STEP_ORDER';
export const ADD_FUNNEL_STEP_FILTER = '@query/ADD_FUNNEL_STEP_FILTER';
export const UPDATE_FUNNEL_STEP_FILTER = '@query/UPDATE_FUNNEL_STEP_FILTER';
export const REMOVE_FUNNEL_STEP_FILTER = '@query/REMOVE_FUNNEL_STEP_FILTER';
export const UPDATE_FUNNEL_STEP_TIMEZONE = '@query/UPDATE_FUNNEL_STEP_TIMEZONE';
export const ADD_FILTER = '@query/ADD_FILTER';
export const UPDATE_FILTER = '@query/UPDATE_FILTER';
export const REMOVE_FILTER = '@query/REMOVE_FILTER';

export const DEFAULT_ANALYSIS = 'count';
export const DEFAULT_TIMEFRAME = 'this_14_days';
export const DEFAULT_TIMEZONE = 'UTC';
export const DEFAULT_EXTRACTION_LIMIT = 100;

export const DEFAULT_FILTER = {
  propertyName: undefined,
  operator: undefined,
  propertyValue: undefined,
  propertyType: undefined,
};

export const DEFAULT_FUNNEL_STEP: FunnelStep = {
  eventCollection: undefined,
  actorProperty: undefined,
  timeframe: DEFAULT_TIMEFRAME,
  timezone: DEFAULT_TIMEZONE,
  filters: [],
  inverted: false,
  optional: false,
  withActors: false,
};
