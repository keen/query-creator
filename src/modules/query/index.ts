import { queryReducer, initialState } from './reducer';
import {
  getQuery,
  getEventCollection,
  getAnalysis,
  getPercentile,
  getTargetProperty,
  getTimeframe,
  getTimezone,
  getGroupBy,
  getOrderBy,
  getInterval,
  getLimit,
  getFunnelSteps,
  getExtractionLimit,
  getExtractionPropertyNames,
  getFilters,
} from './selectors';

import {
  addFilter,
  removeFilter,
  updateFilter,
  setQuery,
  serializeQuery,
  resetQuery,
  selectEventCollection,
  selectTargetProperty,
  selectTimezone,
  selectAnalysis,
  setPercentile,
  setGroupBy,
  setOrderBy,
  setInterval,
  setLimit,
  setExtractionLimit,
  setPropertyNames,
  setTimeframe,
  setFilters,
  setFunnelStepFilters,
  setFunnelSteps,
  addFunnelStep,
  selectFunnelStepCollection,
  updateFunnelStep,
  removeFunnelStep,
  changeFunnelStepsOrder,
  cloneFunnelStep,
  addFunnelStepFilter,
  updateFunnelStepFilter,
  removeFunnelStepFilter,
  resetExtraction,
} from './actions';

import {
  SET_QUERY,
  SET_GROUP_BY,
  SELECT_EVENT_COLLECTION,
  SELECT_TIMEZONE,
  SELECT_FUNNEL_STEP_EVENT_COLLECTION,
  DEFAULT_TIMEZONE,
  DEFAULT_TIMEFRAME,
  DEFAULT_FUNNEL_STEP,
  DEFAULT_FILTER,
  SERIALIZE_QUERY,
  UPDATE_FUNNEL_STEP_TIMEZONE,
} from './constants';
import {
  ReducerState,
  InitialQuery,
  SetQueryAction,
  SerializeQueryAction,
  SelectTimezoneAction,
  SelectFunnelStepEventCollectionAction,
  SelectEventCollectionAction,
  UpdateFunnelStepTimezoneAction,
} from './types';

export {
  initialState,
  queryReducer,
  getQuery,
  getPercentile,
  getFunnelSteps,
  getEventCollection,
  getTargetProperty,
  getExtractionLimit,
  getExtractionPropertyNames,
  getAnalysis,
  getTimezone,
  getGroupBy,
  getOrderBy,
  getLimit,
  getInterval,
  getTimeframe,
  getFilters,
  setTimeframe,
  setGroupBy,
  setOrderBy,
  setLimit,
  setInterval,
  setExtractionLimit,
  setQuery,
  serializeQuery,
  setPropertyNames,
  setPercentile,
  setFilters,
  selectTimezone,
  addFunnelStep,
  setFunnelSteps,
  setFunnelStepFilters,
  updateFunnelStep,
  removeFunnelStep,
  selectFunnelStepCollection,
  changeFunnelStepsOrder,
  cloneFunnelStep,
  addFunnelStepFilter,
  updateFunnelStepFilter,
  removeFunnelStepFilter,
  selectEventCollection,
  selectTargetProperty,
  selectAnalysis,
  resetExtraction,
  resetQuery,
  addFilter,
  updateFilter,
  removeFilter,
};
export {
  InitialQuery,
  SetQueryAction,
  SerializeQueryAction,
  SelectTimezoneAction,
  SelectEventCollectionAction,
  SelectFunnelStepEventCollectionAction,
  ReducerState,
  UpdateFunnelStepTimezoneAction,
};

export {
  SERIALIZE_QUERY,
  SET_QUERY,
  SET_GROUP_BY,
  SELECT_EVENT_COLLECTION,
  SELECT_TIMEZONE,
  SELECT_FUNNEL_STEP_EVENT_COLLECTION,
  DEFAULT_TIMEFRAME,
  DEFAULT_TIMEZONE,
  DEFAULT_FUNNEL_STEP,
  DEFAULT_FILTER,
  UPDATE_FUNNEL_STEP_TIMEZONE,
};
