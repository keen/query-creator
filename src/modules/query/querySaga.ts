import { takeLatest } from 'redux-saga/effects';

import { selectTimezone, updateFunnelStepTimezone } from './actions';
import { updateTimeframeOffset, updateFunnelStepTimeframeOffset } from './saga';

export function* querySaga() {
  yield takeLatest(selectTimezone.type, updateTimeframeOffset);
  yield takeLatest(
    updateFunnelStepTimezone.type,
    updateFunnelStepTimeframeOffset
  );
}
