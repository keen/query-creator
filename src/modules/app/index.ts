import {
  appStart,
  storeSchemas,
  setQueryReadiness,
  fetchProjectDetails,
  fetchProjectDetailsError,
  fetchProjectDetailsSuccess,
} from './actions';
import { appReducer } from './reducer';
import { getQueryReadiness } from './selectors';
import { ReducerState } from './types';
import { APP_START, FETCH_PROJECT_DETAILS } from './constants';

export {
  appStart,
  appReducer,
  setQueryReadiness,
  storeSchemas,
  fetchProjectDetails,
  fetchProjectDetailsError,
  fetchProjectDetailsSuccess,
  getQueryReadiness,
  ReducerState,
};
export { APP_START, FETCH_PROJECT_DETAILS };
