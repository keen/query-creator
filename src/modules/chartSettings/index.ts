import { initialState, chartSettingsReducer } from './reducer';
import { updateChartSettings, resetChartSettings } from './actions';

import { getChartSettings } from './selectors';

import { UPDATE_CHART_SETTINGS, RESET_CHART_SETTINGS } from './constants';

import { ChartSettingsActions, ReducerState } from './types';

export {
  initialState,
  chartSettingsReducer,
  updateChartSettings,
  resetChartSettings,
  getChartSettings,
  UPDATE_CHART_SETTINGS,
  RESET_CHART_SETTINGS,
  ChartSettingsActions,
  ReducerState,
};
