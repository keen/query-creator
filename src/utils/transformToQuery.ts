/* eslint-disable @typescript-eslint/no-unused-vars */
import snakeCase from 'snakecase-keys';

import { convertAbstractOperators } from './convertAbstractOperators';

import { ReducerState as QueryState, InitialQuery } from '../modules/query';

import { FIELDS_CONFIG } from '../config';

export const transformToQuery = (query: QueryState) => {
  const { analysisType, ...queryProperties } = query;
  const queryCopy = { ...queryProperties };

  Object.keys(queryProperties).forEach((fieldName) => {
    const fieldConfig = FIELDS_CONFIG[fieldName];
    if (fieldConfig && !fieldConfig.includes(analysisType)) {
      delete queryCopy[fieldName];
    }
  });

  const {
    filters,
    orderBy,
    propertyNames,
    steps,
    ...restQuerySettings
  } = queryCopy;
  const queryToTransform: InitialQuery = {
    analysisType,
    ...restQuerySettings,
  };

  if (filters) {
    queryToTransform.filters = filters.map(({ id, ...filter }) =>
      convertAbstractOperators(filter)
    );
  }

  if (orderBy) {
    queryToTransform.orderBy = orderBy.map(({ id, ...item }) => item);
  }

  if (propertyNames) {
    queryToTransform.propertyNames = queryCopy.propertyNames.map(
      ({ id, propertyName }) => propertyName
    );
  }

  if (steps) {
    queryToTransform.steps = steps.map(({ id, ...step }) => ({
      ...step,
      filters: step.filters.map(({ id, ...filter }) =>
        convertAbstractOperators(filter)
      ),
    }));
  }

  return snakeCase(queryToTransform);
};
