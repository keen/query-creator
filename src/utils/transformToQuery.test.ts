import { transformToQuery } from './transformToQuery';
import { ReducerState } from '../modules/query';

test('transforms query', () => {
  const query = {
    eventCollection: 'logins',
    targetProperty: null,
    timezone: 'UTC',
    groupBy: ['user.age'],
    orderBy: [
      {
        id: 'a5c959be-dabb-4198-b440-fc43902beaf1',
        propertyName: 'user.age',
        direction: 'ASC',
      },
    ],
    timeframe: 'this_14_days',
    analysisType: 'count',
    steps: [],
    filters: [],
  } as ReducerState;

  expect(transformToQuery(query)).toMatchSnapshot();
});
