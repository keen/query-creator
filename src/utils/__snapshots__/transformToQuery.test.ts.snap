// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms query 1`] = `
Object {
  "analysis_type": "count",
  "event_collection": "logins",
  "filters": Array [],
  "group_by": Array [
    "user.age",
  ],
  "order_by": Array [
    Object {
      "direction": "ASC",
      "property_name": "user.age",
    },
  ],
  "timeframe": "this_14_days",
  "timezone": "UTC",
}
`;
